Make `persist` resettable and fix frame shape mismatch error in BoT-SORT

Y-T-G opened the pull request:
**Minimal Reproducible Example**
```python
>>> from ultralytics import YOLO, ASSETS
>>> model = YOLO("yolov8n.pt")
>>> results = model.track(ASSETS / "zidane.jpg", persist=True)

image 1/1 /ultralytics/ultralytics/assets/zidane.jpg: 384x640 2 persons, 7.6ms
Speed: 1.1ms preprocess, 7.6ms inference, 1.4ms postprocess per image at shape (1, 3, 384, 640)
>>> results = model.track(ASSETS / "bus.jpg", persist=True)

Traceback (most recent call last):
  Cell In[23], line 1
    results = model.track(ASSETS / "bus.jpg", persist=True)
  File /ultralytics/ultralytics/engine/model.py:601 in track
    return self.predict(source=source, stream=stream, **kwargs)
  File /ultralytics/ultralytics/engine/model.py:554 in predict
    return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)
  File /ultralytics/ultralytics/engine/predictor.py:168 in __call__
    return list(self.stream_inference(source, model, *args, **kwargs))  # merge list of Result into one
  File /opt/conda/lib/python3.10/site-packages/torch/utils/_contextlib.py:36 in generator_context
    response = gen.send(None)
  File /ultralytics/ultralytics/engine/predictor.py:262 in stream_inference
    self.run_callbacks("on_predict_postprocess_end")
  File /ultralytics/ultralytics/engine/predictor.py:399 in run_callbacks
    callback(self)
  File /ultralytics/ultralytics/trackers/track.py:80 in on_predict_postprocess_end
    tracks = tracker.update(det, im0s[i])
  File /ultralytics/ultralytics/trackers/byte_tracker.py:333 in update
    warp = self.gmc.apply(img, dets)
  File /ultralytics/ultralytics/trackers/utils/gmc.py:115 in apply
    return self.applySparseOptFlow(raw_frame)
  File /ultralytics/ultralytics/trackers/utils/gmc.py:343 in applySparseOptFlow
    matchedKeypoints, status, _ = cv2.calcOpticalFlowPyrLK(self.prevFrame, frame, self.prevKeyPoints, None)
error: OpenCV(4.10.0) /io/opencv/modules/video/src/lkpyramid.cpp:1394: error: (-215:Assertion failed) prevPyr[level * lvlStep1].size() == nextPyr[level * lvlStep2].size() in function 'calc'
```

The PR makes `persist` changeable after the first prediction. Users can set it to `False` after having it `True` prior in order to manually reset the trackers. In case the user doesn't reset it, the error is prevented by checking for frame shapes and resetting the "GMC" state with a warning.

**Without setting `persist=False` before inferring on a frame with different shape causes warning**
```python
>>> results = model.track(ASSETS / "bus.jpg", persist=True)

image 1/1 /ultralytics/ultralytics/assets/bus.jpg: 640x480 3 persons, 1 bus, 7.5ms
Speed: 1.8ms preprocess, 7.5ms inference, 1.3ms postprocess per image at shape (1, 3, 640, 480)
>>> results = model.track(ASSETS / "zidane.jpg", persist=True)

WARNING ⚠️ Current frame has a different shape from previous frame. If you're tracking a new video, set persist=False for the first frame of the new video to reset the tracker.
image 1/1 /ultralytics/ultralytics/assets/zidane.jpg: 384x640 1 person, 7.8ms
Speed: 1.2ms preprocess, 7.8ms inference, 1.2ms postprocess per image at shape (1, 3, 384, 640)
```

**Setting `persist=False` before inferring on a frame with different shape causes tracker to be reset and hence no warning**
```python
>>> results = model.track(ASSETS / "bus.jpg", persist=True)

image 1/1 /ultralytics/ultralytics/assets/bus.jpg: 640x480 3 persons, 1 bus, 6.8ms
Speed: 1.9ms preprocess, 6.8ms inference, 1.3ms postprocess per image at shape (1, 3, 640, 480)
>>> results = model.track(ASSETS / "zidane.jpg", persist=False)

image 1/1 /ultralytics/ultralytics/assets/zidane.jpg: 384x640 2 persons, 7.4ms
Speed: 1.2ms preprocess, 7.4ms inference, 1.2ms postprocess per image at shape (1, 3, 384, 640)
```


<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
--->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improves how object tracking persistence is managed, making tracker resets more reliable and user-friendly during video processing. 🛠️🎥

### 📊 Key Changes
- Refactors tracker persistence logic: The `persist` flag is now managed through predictor arguments instead of being passed through multiple functions.
- Simplifies tracker registration: Removes unnecessary use of `functools.partial` and streamlines callback registration.
- Adds a warning for frame shape mismatches: Alerts users if video frames change size unexpectedly, suggesting when to reset the tracker.
- Improves tracker reset handling: Ensures trackers reset correctly when starting a new video or when frame shapes differ.

### 🎯 Purpose & Impact
- 🧹 **Cleaner code**: Reduces complexity by centralizing persistence logic, making the codebase easier to maintain.
- 🎬 **More robust tracking**: Prevents tracking errors when processing multiple videos or videos with changing frame sizes.
- 🚨 **Better user feedback**: Warns users about potential issues and provides guidance on how to avoid tracking problems.
- 👍 **Improved reliability**: Ensures object tracking works as expected across different video scenarios, benefiting both developers and end users.

Y-T-G committed:
Make `persist` resettable and warn on frame shape mismatch

UltralyticsAssistant commented:
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to ensure a smooth integration of your proposed changes. Please review the checklist below to facilitate the process:

- ✅ **Purpose and Clarity**: Clearly explain the purpose of this PR and how it addresses the error with frame shape mismatches while using `botsort`. Check if this relates to any existing [issues](https://github.com/ultralytics/ultralytics/issues) and reference them in your PR description. Ensure commit messages are clear and adhere to the project's conventions.
- ✅ **Stay Updated**: Ensure your PR is in sync with the `ultralytics/ultralytics` `main` branch. If there are conflicts, resolve them by clicking the 'Update branch' button or via `git pull` and `git merge main` in your local environment.
- ✅ **Pass CI Checks**: Confirm that all Ultralytics [CI checks](https://docs.ultralytics.com/help/CI/) are passing. In case of any failures, please address them promptly.
- ✅ **Documentation Needs**: If you've introduced or modified features, update the relevant [documentation](https://docs.ultralytics.com) to reflect these changes.
- ✅ **Testing**: Add or update tests as necessary to cover your changes, ensuring all tests pass successfully. A minimum reproducible example (MRE) is appreciated and appears to be included! 🎉
- ✅ **CLA Signature**: Ensure you have signed our [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/) if this is your first contribution. Leave a comment like "I have read the CLA Document and I sign the CLA" if needed.
- ✅ **Efficient Code**: Keep changes to the minimum necessary to resolve the issue. As Bruce Lee says, _"Hack away the unessential."_ 🥋

For further guidance, refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Feel free to ask any questions by leaving a comment. An Ultralytics engineer will assist you soon. Thank you for contributing to Ultralytics! 🚀✨

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Y-T-G committed:
Update gmc.py

glenn-jocher committed:
Update CI.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q commented:
@Y-T-G Thanks for the PR! I think we prefer not to include `persist` in defualt.yaml though, is there a way to add this warning and keep the `persist` as it is?

glenn-jocher committed:
Merge branch 'main' into persist_fix

UltralyticsAssistant committed:
Merge branch 'main' into persist_fix

UltralyticsAssistant committed:
Merge branch 'main' into persist_fix

UltralyticsAssistant committed:
Merge branch 'main' into persist_fix

UltralyticsAssistant committed:
Merge branch 'main' into persist_fix

Y-T-G committed:
remove from default.yaml

Y-T-G committed:
Merge branch 'main' into persist_fix

Y-T-G committed:
Merge branch 'main' into persist_fix

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Y-T-G commented:
@Laughing-q I updated it so that `predictor.args.persist` is now updated everytime `track()` is called. That way it's resettable and not fixed once the tracker is registered.

Y-T-G committed:
check None

Y-T-G committed:
Pass on register

Y-T-G committed:
add persist to args on initialization

Y-T-G committed:
remove warning prefix

Y-T-G committed:
Update model.py

Laughing-q commented:
@Y-T-G since we are adding tracker in callbacks, can we use [model.reset_callbacks](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/model.py#L1008-L1008) here to remove registered callbacks to reset tracker?

Y-T-G commented:
@Laughing-q Reset it when `persist` is False inside `model.track()`?

UltralyticsAssistant committed:
Merge branch 'main' into persist_fix

Y-T-G committed:
Merge branch 'main' into persist_fix

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Merge main into persist_fix

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Merge branch 'main' into persist_fix

glenn-jocher commented:
Hi @Y-T-G, thanks for this one! It's been open since October 2024 and the tracker code has moved substantially since — rather than resurrect a two-year-old branch, any part of this still relevant (persist resetting, frame-shape handling on source changes) is best re-proposed fresh against current main. Closing as too old. Thank you for all your contributions!
